### PR TITLE
Upgrade to Lit 2

### LINKED ISSUE #85: Lit 2 version
Please create a version 2.0.0(.pre1?) that uses Lit 2 (currently `2.0.0.pre1`)

### PR DESCRIPTION
**Pull request by Artur-**
## Description

Uses the latest available Lit 2 pre release

## Related Issue

Fixes #85

## Motivation and Context

Makes it possible to test and use lit-mobx with Lit 2

## How Has This Been Tested?

Existing tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: test/lit-mobx.test.ts [hunk 1]
```diff
@@ -24,9 +24,11 @@ import './test-elements';
 
 describe('MobxLitElement', () => {
     it('updates arrays', async () => {
-        const el = await fixture<TestMobxLitElementUpdatesArrays>(html`
-            <test-mobx-lit-element-updates-arrays></test-mobx-lit-element-updates-arrays>
-        `);
+        const el = await fixture<TestMobxLitElementUpdatesArrays>(
```
Comment (Westbrook) [inline test/lit-mobx.test.ts]: I'm half surprised this work. In _theory_, none of these tests use special bindings meaning if we use the `html` from `@open-wc/testing` in this file we should be able to remove the `as any` casting.

**Review by Westbrook — COMMENTED** [review]
@benjamind we've gotten away without too many maintenance relative changes to this to date, so it might not be necessary, but do you have any negative thoughts about _only_ supporting `lit@2` here, and not `lit@1` as well? With the low amount of changes to our library, so far, simply requiring users use one version of the other is not that big of a deal, we do a major version bump and get over it. I certainly hope for a quick ecosystem change to `lit@2`, but if we needed to go back and do something for `lit@1` later on, do you prefer maintaining that work from a separate branch or doing the work with `||` operators and export maps to support the two versions side by side?

File: test/lit-mobx.test.ts (hunk 1 above)
Comment (Artur-) [reply test/lit-mobx.test.ts]: Oh I forgot to comment that the tests do not work with this PR. I did not test using `html` from open-wc though. I imagine that open-wc also will make a lit 2 version ”soon enough” so I pushed what I had here as a draft

Comment (Artur-) [reply test/lit-mobx.test.ts]:
```
> tsc --build tsconfig.json && tsc --build demo/tsconfig.json

node_modules/@lit/reactive-element/reactive-element.d.ts:24:11 - error TS2451: Cannot redeclare block-scoped variable 'JSCompiler_renameProperty'.

24     const JSCompiler_renameProperty: <P extends PropertyKey>(prop: P, _obj: unknown) => P;
             ~~~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/lit-element/lib/updating-element.d.ts:15:9
    15     var JSCompiler_renameProperty: <P extends PropertyKey>(prop: P, _obj: unknown) => P;
               ~~~~~~~~~~~~~~~~~~~~~~~~~
    'JSCompiler_renameProperty' was also declared here.
```

Comment (Westbrook) [reply test/lit-mobx.test.ts]: Can you try with `html` from there and see what you get? Being we're in NPM with no resolutions, I think it should would, and be resilient to both sides of the `lit@2` release.

Comment (Artur-) [reply test/lit-mobx.test.ts]: The `JSCompiler_renameProperty` problem still blocks compilation of the tests

**Comment by benjamind** [comment]
> @benjamind we've gotten away without too many maintenance relative changes to this to date, so it might not be necessary, but do you have any negative thoughts about _only_ supporting `lit@2` here, and not `lit@1` as well? With the low amount of changes to our library, so far, simply requiring users use one version of the other is not that big of a deal, we do a major version bump and get over it. I certainly hope for a quick ecosystem change to `lit@2`, but if we needed to go back and do something for `lit@1` later on, do you prefer maintaining that work from a separate branch or doing the work with `||` operators and export maps to support the two versions side by side?

Good question, I guess a v1 for lit@1 and a v2 for lit@2 seems reasonable to me. I had wondered if with lit@2 we might want to consider moving away from the mixin pattern and towards the reactive controller pattern instead, but I've not had the time to look into this yet.

Export maps is definitely an option, but I'm not entirely up to speed on how good support for this is at this time?

**Comment by benjamind** [comment]
Thanks @Artur- for this! I'll try and find time to dig into it asap.

File: test/lit-mobx.test.ts (hunk 1 above)
Comment (Artur-) [reply test/lit-mobx.test.ts]: I created https://github.com/Polymer/lit-html/issues/1655 for the JSCompiler error

**Comment by Artur-** [comment]
I published this as `a-lit-mobx@2.0.0-pre.1` to be able to test in other projects

**Comment by benjamind** [comment]
I had a discussion with @justinfagnani on best approach for lit2 support. We concluded that the mixin approach remains the best approach for mobx due to the way mobx works we want to always wrap the entire render function. Controllers would offer more fine grained control, but this would then make the library harder to use.

With regard to how to vend a version which supports both v1 and v2 lit, conversations are still ongoing as to the best approach.

**Comment by Artur-** [comment]
The work that needs to be done is drastically reduced with Lit 2.0.0-pre.2. I updated this PR with the minimum that contains no code changes at all.

As no code changes are needed, it seems like there could be a `"lit-element": "^2.0.0 || ^3.0.0"` dependency instead if you want to stay compatible with both LitElement versions

**Comment by benjamind** [comment]
Excellent.

In discussion. With @justinfagnani we concluded we should probably make two test projects using the different major versions of lit-element to confirm ongoing compatibility, but agree also with your assessment here that we should just be able to allow both versions to be compatible now.

Thanks for staying on top of this @Artur- !

**Comment by Artur-** [comment]
Updated to Lit 2 rc1. 

What's the plan here? New major? Same major? Wait for Lit 2 final? Release with rc1?

We will start asap moving all our projects to Lit 2 and I would prefer not to use a fork for that.

**Comment by benjamind** [comment]
We're planning a similar shift ASAP, my current thinking is we release this under new major?

**Comment by Westbrook** [comment]
I'd vote that in this case we can require a new major and have it support only Lit@2 going forward. Saves space to make bug fixes if absolutely needed, but saves us the work of having to think too much about splitting support between the two versions in a single release.

**Comment by jordanaustin** [comment]
Is the plan to wait until a stable Lit release before moving forward with this change, or do you feel good about basing this off the RC.1? Selfishly I'd love to see this land as soon as possible 😁

**Comment by benjamind** [comment]
I believe the semver `^2.0.0` should incorporate the rc verions too so maybe we can just go ahead? I'd like to get the tests updated to use the new lit version though before we release. I'm juggling a few things right now so haven't had a chance to do this yet.

**Comment by jordanaustin** [comment]
> I believe the semver `^2.0.0` should incorporate the rc verions too so maybe we can just go ahead? I'd like to get the tests updated to use the new lit version though before we release. I'm juggling a few things right now so haven't had a chance to do this yet.

After looking at updating the tests, it looks like `fixture` from `@open-wc/testing` may need to be updated before we can do this. But I haven't looked to deep at this yet

**Comment by benjamind** [comment]
Yup confirmed, taking a look at this today and will see what we can do.

**Comment by benjamind** [comment]
Have put together a full featured PR here if you want to take a look - https://github.com/adobe/lit-mobx/pull/93

**Comment by Artur-** [comment]
#93 looks good and more complete so let's close this then

**Pull request closed without merging**